### PR TITLE
HRINT-2432  [Demo] Queries: Faceted Search

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

**Comment by Danielle9897** [comment]
@DejanMilicic can you pls review ?

**Review by DejanMilicic — CHANGES_REQUESTED** [review]

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsAggregations/FacetsAggregationsController.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    public class FacetsAggregationsController : DemoCodeController
+    {
+        public FacetsAggregationsController(UserIdContainer userId, UserStoreCache userStoreCache, MediaStoreCache mediaStoreCache,
+            DatabaseSetup databaseSetup) : base(userId, userStoreCache, mediaStoreCache, databaseSetup)
+        {
+        }
+
+        #region Demo
+        #region Step_1
+        public class Products_ByCategoryPriceAndUnits : AbstractIndexCreationTask<Product, Products_ByCategoryPriceAndUnits.IndexEntry>
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsAggregations/FacetsAggregationsController.cs:27]: No specific conventions that I know about...

If I look for example at an **auto-index** generated it looks like this:
`Auto/Companies/ByContactAndExternalIdAndFaxAndNameAndPhone`

I wonder what is best for this **static index:**
1. Products/ByCategoryPriceAndUnits
2. Products/ByCategory/ByPrice/ByUnits
3. Products/ByCategoryAndPriceAndUnits
4. Products/ByCategory/Price/Units

?

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsBasic/metadata.json [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+                    "Url"  : "/faceted-search/facets-aggregations"
+                }
+            ]
+        },
+        {
+            "Title": "Define a RangeFacet",
+            "Slug": "step-4",
+            "DescriptionLines": [
+                "* _Ranges_ - Define ranges on an index entry field by which the documents will be grouped.<br>",
+                "  A separated aggregated item will be created per range defined.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsBasic/metadata.json:149]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                },
+                {
+                    "Title": "Index Related Documents Demo",
+                    "Type" : "demo",
+                    "Url"  : "/related-documents/index-related-documents"
+                }
+            ]
+        },
+        {
+            "Title": "Define class FacetSetup",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:76]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                },
+                {
+                    "Title": "Storing Facets Example",
+                    "Type" : "link",
+                    "Url"  : "https://ravendb.net/docs/article-page/latest/Csharp/client-api/session/querying/how-to-perform-a-faceted-search#example-iii"
+                }
+            ]
+        },
+        {
+            "Title": "Store the Facet document",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:106]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            "Slug": "step-3",
+            "DescriptionLines": [
+                "* Store the 'facetSetup' in the session and save changes.",
+                "",
+                "* The setup will be stored in a document with the specified ID under the `FacetSetup` collection."
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:111]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                },
+                {
+                    "Title": "Storing Facets Example",
+                    "Type" : "link",
+                    "Url"  : "https://ravendb.net/docs/article-page/latest/Csharp/client-api/session/querying/how-to-perform-a-faceted-search#example-iii"
+                }
+            ]
+        },
+        {
+            "Title": "execute the query",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:132]: Fixed

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                }
+            ]
+        },
+        {
+            "Title": "Define class FacetSetup",
+            "Slug": "step-2",
+            "DescriptionLines": [
+                "Define class `FacetSetup`:",
+                "",
+                "* _Id_ - Provide the document ID in which the facet setup will be stored",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:81]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                    "Url"  : "https://ravendb.net/docs/article-page/latest/Csharp/client-api/session/querying/how-to-perform-a-faceted-search#example-iii"
+                }
+            ]
+        },
+        {
+            "Title": "execute the query",
+            "Slug": "step-4",
+            "DescriptionLines": [
+                "* Define a query on the 'Products' collection using the defined index.<br>",
+                "  Pass the facets document ID to the `AggregateUsing()` method and execute the query.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/metadata.json:136]: done

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/FacetsFromDocumentController.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            
+            #region Demo
+            Dictionary<string, FacetResult> queryResults;
+
+            using (IDocumentSession session = DocumentStoreHolder.Store.OpenSession())
+            {
+                #region Step_2
+                FacetSetup facetSetup = new FacetSetup
+                {   
+                    Id = "myFacetSetupDocumentID",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsFromDocument/FacetsFromDocumentController.cs:67]: image

File: DemoServer/Controllers/Demos/FacetedSearch/FacetsAggregations/FacetsAggregationsController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DejanMilicic) [reply DemoServer/Controllers/Demos/FacetedSearch/FacetsAggregations/FacetsAggregationsController.cs:27]: Let's leave it like it is

**Review by DejanMilicic — APPROVED** [review]

**Pull request merged**